### PR TITLE
[Semi-Auto] Add cross_entropy_with_softmax infer_backward rule

### PR DESCRIPTION
**Pull request by pkuzyc**
### PR types
Function optimization

### PR changes
Others

### Description
Pcard-70448
Add infer_backward rule for cross_entropy_with_softmax to infer inputs' dims mappings from outputs'.

### CODE REVIEW

File: paddle/fluid/distributed/auto_parallel/spmd_rules/cross_entropy_with_softmax_spmd_rule.cc [hunk 1]
```diff
@@ -173,10 +173,116 @@ CrossEntropyWithSoftmaxSPMDRule::InferForward(
...
+
+  // normalize axis
+  if (axis < 0) {
+    axis = x_ndim + axis;
+  }
+
+  std::string alphabet =
+      "abcdefghijlmnopqrstuvwxyz";  // k for softmax_normalize axis
+  std::string x_axes = GetBroadcastAxes(x_ndim, x_ndim, alphabet);
+  x_axes[axis] = 'k';
```
Comment (JZ-LIANG) [inline paddle/fluid/distributed/auto_parallel/spmd_rules/cross_entropy_with_softmax_spmd_rule.cc:214]: should check if softmax_normalize is the last axis, if it is not, it could not be sharded.

and if softmax_normalize is the last axis,but if soft_label == true,  it could not be sharded.

File: test/auto_parallel/spmd_rules/test_cross_entropy_with_softmax_rule.py [hunk 2]
```diff
@@ -147,6 +150,122 @@ def test_cross_entropy_with_softmax_infer_forward(self):
...
+
+        self.assertEqual(
+            infered_output_dist_attrs[0].dims_mapping, [1, 0, -1]
+        )  # softmax output
+        self.assertEqual(
+            infered_output_dist_attrs[1].dims_mapping, [1, 0, -1]
+        )  # loss
+
+        # GPT MP case, shard normalized axis
+        # [-1, -1, 0], [-1, -1, -1] (outputs) -->
```
Comment (JZ-LIANG) [inline test/auto_parallel/spmd_rules/test_cross_entropy_with_softmax_rule.py:187]: distinguish 
outputs: loss (last axis is 1) & softmax_out
inputs: label(last axis maybe 1)  & logits

**Review by JZ-LIANG — APPROVED** [review]
Update in Next PR
